Update Add Authentication to Your Vanilla JavaScript App in 20 Minutes (#1020)
oktadeveloper/okta-blog: _source/_posts/2018-06-05-authentication-vanilla-js.md

 - "Want to get your hands dirty with good ol' JavaScript? This tutorial doesn't use any JavaScript frameworks to build an app. The build process is amazing! 🤘"
type: conversion
github: https://github.com/oktadev/okta-vanilla-js-example
changelog:
  - 2020-01-06: Updated dependencies, code, and screenshots. See this post's changes in [okta-blog#1020](https://github.com/oktadev/okta-blog/pull/1020) and the example app changes in [okta-vanilla-js-example#2](https://github.com/oktadev/okta-vanilla-js-example/pull/2). 
---

*"Sometimes nothing is good enough"* is a phrase that software engineers don't speak or hear often. In the fast-changing world of web development, there is no shortage of bleeding-edge JavaScript frameworks promising to make your life easier or inch out its predecessors. You may ask yourself if it is even possible to build a modern web application without one of these frameworks, let alone add secure authentication. Well, it is!

Vanilla JavaScript is frequently used to describe ordinary JavaScript leveraging the browser and DOM APIs. Yes, you read that correctly. No Angular, React, Vue, or even jQuery. And even better, it's not as complicated as you might assume.

> There are many reasons you may decide to forgo the framework bloat for vanilla JavaScript. Its raw speed and non-complicated build process are among my top reasons. Regardless of your reasons, this article might serve as a refresher on leveraging the power of JavaScript to create some great user experiences.

Reviewer: ```suggestion
There are many reasons you may decide to forgo the framework bloat for vanilla JavaScript. Among my top reasons: raw speed and noncomplicated build process. Regardless of your reasons, this article serves as a refresher on leveraging the power of JavaScript to create some great user experiences.
```
makes it even simpler and more direct

### Test Out Your New Vanilla JS App

> Congrats! You have successfully built a dynamic SPA in around 100 lines of code. Not bad! Go back to your browser and hit refresh. You should initially see no meals as the server starts with an empty array. You should add meals, remove meals, and refresh the page to see the app in action.

Reviewer: ```suggestion
Congrats! You have successfully built a dynamic SPA in around 100 lines of code. Not bad! Go back to your browser and hit refresh. You should initially see no meals as the server starts with an empty array. To see the app in action you can add meals or remove meals, then refresh the page to view the changes.
```
clarifying and varying the sentences so the the "You..." is less repetitive

## Add Authentication with Okta's Sign-In Widget

[Okta's Sign-in Widget](https://developer.okta.com/code/javascript/okta_sign-in_widget) is a JavaScript library that gives you a full-featured and customizable login widget that can be added to any website.  With just a few lines of code, you can implement a login flow to your app.

### Configure a New Okta application

To use the Okta Sign-in Widget, you must first have an Okta developer account. If you don't have one you can [create a free account](https://developer.okta.com/signup/). After you are logged in, click on the **Applications** tab and then **Create App Integration** button. 

> Select **OIDC - OpenID Connect** -> **Single-Page Application** and click on the **Next** button. Select **Implicit (hybrid)** as the **Grant type** and un-check **Authorization Code**. Now, Enter `http://localhost:8080/` as your **Sign-in redirect URIs**. Finally, select **Allow everyone in your organization to access** as the **Controlled access** and click on the **Save** button. 

Reviewer 2: You should be able to replace these two paragraphs with a CLI include. Examples at https://developer.okta.com/blog/setup.

Your Application settings should look similar to this:

> {% img blog/vanilla-js/app-settings.png alt:"SPA App Settings" width:"600" %}{: .center-image }

Reviewer 2: I don't think this screenshot is necessary. We try not to include them anymore since our whole UI changed in February 2021 and we don't want to have to update all posts when the UI changes the next time. I've also noticed none of our docs include screenshots.

Reviewer 2: Related: This post should use auth code flow with PKCE. We try not to use implicit flow except when absolutely necessary.

### Enable CORS

> Now, [enable CORS](https://developer.okta.com/docs/guides/enable-cors/main/) for `http://localhost:8080/` in your Okta Admin Dashboard.

Reviewer 2: I don't think this part is necessary when using the CLI. You might mention that it's for folks that used the UI and not the CLI (e.g., If you used the Admin Console to create your app). Otherwise, you might not need it.

Author: Ah, that's cool. I did not realize that.

    this.signIn = new OktaSignIn({
      baseUrl: 'https://{yourOktaDomain}',
      clientId: '{clientId}',

>       redirectUri: 'http://localhost:8080',

Reviewer 2: Any particular reason you changed this value? `window.location.origin` should work and it's more dynamic.

      authParams: {
        issuer: 'https://{yourOktaDomain}/oauth2/default',

>         pkce: false,

Reviewer 2: Why false? You can use PKCE with the sign-in widget.

### Initialize the Okta Sign-in Widget

To initialize the widget, you instantiate a new `OktaSignIn` object.  Remember to change the `{yourOktaDomain}` (twice) and `{clientId}` to yours.

```js
this.signIn = new OktaSignIn({
  baseUrl: 'https://{yourOktaDomain}',
  clientId: '{clientId}',

>   redirectUri: 'http://localhost:8080',

Reviewer 2: Same comment as above.

}
```

> You can check out [https://ok1static.oktacdn.com/assets/js/sdk/okta-signin-widget/2.6.0/css/okta-theme.css](https://ok1static.oktacdn.com/assets/js/sdk/okta-signin-widget/2.6.0/css/okta-theme.css) for a complete list of CSS classes you can inject your styles.

Reviewer: ```suggestion
To inject your own styles, check out [this complete list of CSS classes](https://ok1static.oktacdn.com/assets/js/sdk/okta-signin-widget/2.6.0/css/okta-theme.css).
```
clarity and avoids URL repetition.
